---
title: "cp_1_preprocessing"
author: "Bonnie EJ Maven"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = FALSE)
```

# 0. Overview

This script preprocesses single-cell RNA-seq (scRNA-seq) data for day 8 cardiac progenitors (CP) output from CellRanger. The workflow includes loading the dataset, creating a Seurat object, and processing the object with sctransform, harmony, etc. This prepares the data for downstream processing and analysis.

# 1. Set up; Load Libraries & Set Working Directory
Load required libraries for analyses.

```{r}
# Load libraries.
library(Seurat)
library(harmony)
library(sctransform)

# Set working directory based on the path of the current open file.
setwd(dirname(rstudioapi::getActiveDocumentContext()$path))
```

# 2. Set up; Load Data
Load in data. A GEO link will be provided upon publication.

```{r}
# Load the seurat_dataset dataset.
seurat_dataset.data <- Read10X(data.dir = "../data/filtered_feature_bc_matrix")

# Assign a base name for folder and file naming consistency.
baseName = "cp_all"

# Create a directory for RDS objects.
dir.create("../data/RDS")

# Create a directory for QC images.
dir.create("../analyses/QC")
```

# 3. Initialize Seurat Object

```{r}
# Initialize the Seurat object with the raw data.
seurat_dataset <- CreateSeuratObject(counts = seurat_dataset.data, project = "D8_CP_all", min.cells = 3, min.features = 200)

# Print an overview of the Seurat object.
seurat_dataset 
# An object of class Seurat
# 28164 features across 57969 samples within 1 assay 
# Active assay: RNA (28164 features, 0 variable features)

# Remove the raw data from the environment.
rm(seurat_dataset.data)

# Set mitochondria features as percent.mt metadata.
seurat_dataset[["percent.mt"]] <- PercentageFeatureSet(seurat_dataset, pattern = "^MT-")

# Add sample names to metadata. Each gem group number corresponds to its order in the original aggregation.csv from CellRanger.
classification.vec <- as.numeric(gsub(".*-","", (colnames(x = seurat_dataset))))
names(classification.vec) <- colnames(x = seurat_dataset)
classification.vec[classification.vec=="1"] <- "WT_CP_R1"
classification.vec[classification.vec=="2"] <- "IKO_CP_R1"
classification.vec[classification.vec=="3"] <- "NKO_CP_R1"
classification.vec[classification.vec=="4"] <- "siRNA_control"
classification.vec[classification.vec=="5"] <- "siRNA_ISL1"
classification.vec[classification.vec=="6"] <- "WT_CP_R2"
classification.vec[classification.vec=="7"] <- "WT_CP_R3"
classification.vec[classification.vec=="8"] <- "IKO_CP_R2"
classification.vec[classification.vec=="9"] <- "IKO_CP_R3"
classification.vec[classification.vec=="10"] <- "NKO_CP_R2"
classification.vec[classification.vec=="11"] <- "NKO_CP_R3"
seurat_dataset$"gem.group" <- classification.vec

# Check that the samples are labeled correctly. 
head(seurat_dataset@meta.data)
```

# 4. Perform QC and Filtering
Apply standard filtering for this cell type.

```{r}
# Visualize and save QC features of the data before further filtering.
pdf(paste("../analyses/QC/", baseName, "_raw_overview.pdf", sep = ""))
VlnPlot(seurat_dataset, features = c("nFeature_RNA", "nCount_RNA", "percent.mt"), ncol = 3)
dev.off()

# 1. Filter cells based on the percentage of mitochondrial genes (percent.mt).
seurat_dataset <- subset(seurat_dataset, subset = percent.mt < 10)

# 2. Filter cells based on the total UMI counts (nCount_RNA).
seurat_dataset <- subset(seurat_dataset, subset = nCount_RNA < 50000)

# 3. Filter cells based on the number of detected genes (nFeature_RNA).
seurat_dataset <- subset(seurat_dataset, subset = nFeature_RNA > 1500 & nFeature_RNA < 6000)

# Visualize and save QC features of the data after filtering.
pdf(paste("../analyses/QC/", baseName, "_filtered_overview.pdf", sep = ""))
VlnPlot(seurat_dataset, features = c("nFeature_RNA", "nCount_RNA", "percent.mt"), ncol = 3)
dev.off()

# Check the Seurat object for an overview after filtering.
seurat_dataset
# An object of class Seurat 
# 28164 features across 30059 samples within 1 assay 
# Active assay: RNA (28164 features, 0 variable features)

# Import cell cycle genes.
cc.genes <- readLines(con = "../data/cell_cycle/cell_cycle_genes.csv")

# Split the list of genes into markers of S phase and markers of G2/M phase.
s.genes <- cc.genes[1:45]
g2m.genes <- cc.genes[46:100]

# Assign cell cycle scores using the identified markers.
seurat_dataset <- CellCycleScoring(seurat_dataset, s.features = s.genes, g2m.features = g2m.genes, set.ident = TRUE)

# Confirm the assignment of cell cycle scores.
head(seurat_dataset@meta.data)

# Clean up the environment.
rm(cc.genes, s.genes, g2m.genes)

# Use SCTransform to normalize and scale data while regressing out cell cycle genes.
seurat_dataset <- SCTransform(seurat_dataset, vars.to.regress = c("S.Score", "G2M.Score"), verbose = TRUE)

# Run ElbowPlot to determine the optimal number of principal components (PCs) for dimensionality reduction.
seurat_dataset <- RunPCA(seurat_dataset, verbose = TRUE)
ElbowPlot(seurat_dataset) # Use 15 dimensions.

# Perform dimensionality reduction and clustering using UMAP with the optimal number of PCs.
seurat_dataset <- RunUMAP(seurat_dataset, dims = 1:15, verbose = TRUE)
seurat_dataset <- FindNeighbors(seurat_dataset, dims = 1:15, verbose = TRUE)
seurat_dataset <- FindClusters(seurat_dataset, verbose = TRUE)

# Save the UMAP plot as part of the QC process.
pdf(paste("../analyses/QC/", baseName, "_filtered_DimPlot.pdf", sep = ""))
DimPlot(seurat_dataset, reduction = "umap", label = T)
DimPlot(seurat_dataset, reduction = "umap", group.by = "gem.group")
dev.off()

# Save the Seurat object before proceeding with batch correction.
saveRDS(seurat_dataset, file = paste("../data/RDS/", baseName, "_filtered.rds", sep = ""))
```

# 5. Batch Correction with Harmony
At this stage, no batch correction has been applied. You can explore the dataset without batch correction, but it is recommended to apply it.

```{r}
# Perform Harmony batch correction using the 'gem.group' variable.
seurat_dataset <- RunHarmony(seurat_dataset, group.by.vars = "gem.group", plot_convergence = T, assay.use = "SCT")

# Run ElbowPlot again to determine the optimal number of principal components (PCs) for dimensionality reduction after batch correction.
seurat_dataset <- RunPCA(seurat_dataset, verbose = TRUE)
ElbowPlot(seurat_dataset) # Use 15 dimensions.

# Perform dimensionality reduction and clustering using UMAP with the optimal number of PCs after batch correction.
seurat_dataset <- RunUMAP(seurat_dataset, dims = 1:15, verbose = TRUE)
seurat_dataset <- FindNeighbors(seurat_dataset, dims = 1:15, verbose = TRUE)
seurat_dataset <- FindClusters(seurat_dataset, verbose = TRUE)

# Save the UMAP plot after batch correction as part of the QC process.
pdf(paste("../analyses/QC/", baseName, "_bc_DimPlot.pdf", sep = ""))
DimPlot(seurat_dataset, reduction = "umap", label = T)
DimPlot(seurat_dataset, reduction = "umap", group.by = "gem.group")
dev.off()

# Save the Seurat object after batch correction (indicated by "bc" in the file name).
saveRDS(seurat_dataset, file = paste("../data/RDS/", baseName, "_bc.rds", sep = ""))

# If you need to return to this point, load the Seurat object with batch correction applied.
seurat_dataset <- readRDS(file = paste("../data/RDS/", baseName, "_bc.rds", sep = ""))
```

# 6. Quick Overview Analysis
This section provides a general check of the data before proceeding with further analyses by visualizing the expression of key cardiac markers and potential contaminants.

```{r}
# Plot UMAPs for MYH6 and TNNT2, two well-known cardiac markers, to visualize their expression across the dataset.
FeaturePlot(seurat_dataset,  features = "MYH6")
FeaturePlot(seurat_dataset,  features = "TNNT2")

# Plot a UMAP for HHEX, an endoderm marker, which is a common contaminant in cardiac differentiation experiments.
FeaturePlot(seurat_dataset,  features = "HHEX")

# Generate a heatmap displaying the expression of various cardiac markers in the dataset.
DoHeatmap(seurat_dataset, features = c("ACTN2", "TNNT2", "NKX2-5", "MYH6", "MYL7", "HAND1"))

# Based on the initial overview, the data appears to be of good quality. Proceed with more in-depth analyses.
```

# 7. sessionInfo
Capture the current session information.

```{r}
sessionInfo()

# To comment out a line of text use the keyboard shortcut: Command + Shift + C


# sessionInfo()
# R version 4.2.0 (2022-04-22)
# Platform: x86_64-apple-darwin17.0 (64-bit)
# Running under: macOS Monterey 12.4
# 
# Matrix products: default
# LAPACK: /Library/Frameworks/R.framework/Versions/4.2/Resources/lib/libRlapack.dylib
# 
# locale:
#   [1] en_US.UTF-8/en_US.UTF-8/en_US.UTF-8/C/en_US.UTF-8/en_US.UTF-8
# 
# attached base packages:
#   [1] grid      stats     graphics  grDevices utils     datasets  methods   base     
# 
# other attached packages:
#   [1] SeuratWrappers_0.3.0  sctransform_0.3.3     shiny_1.7.2           harmony_0.1.0        
# [5] Rcpp_1.0.9            RColorBrewer_1.1-3    circlize_0.4.15       ComplexHeatmap_2.12.0
# [9] cowplot_1.1.1         ggsci_2.9             forcats_0.5.1         stringr_1.4.0        
# [13] purrr_0.3.4           readr_2.1.2           tidyr_1.2.0           tibble_3.1.8         
# [17] ggplot2_3.3.6         tidyverse_1.3.2       patchwork_1.1.1       Seurat_4.1.1         
# [21] sp_1.5-0              SeuratObject_4.1.0    dplyr_1.0.9          
# 
# loaded via a namespace (and not attached):
#   [1] utf8_1.2.2            R.utils_2.12.0        reticulate_1.25      
# [4] tidyselect_1.1.2      htmlwidgets_1.5.4     Rtsne_0.16           
# [7] devtools_2.4.4        munsell_0.5.0         codetools_0.2-18     
# [10] ragg_1.2.2            ica_1.0-3             future_1.27.0        
# [13] miniUI_0.1.1.1        withr_2.5.0           spatstat.random_2.2-0
# [16] colorspace_2.0-3      progressr_0.10.1      rstudioapi_0.13      
# [19] stats4_4.2.0          ROCR_1.0-11           tensor_1.5           
# [22] listenv_0.8.0         labeling_0.4.2        polyclip_1.10-0      
# [25] farver_2.1.1          parallelly_1.32.1     vctrs_0.4.1          
# [28] generics_0.1.3        R6_2.5.1              doParallel_1.0.17    
# [31] clue_0.3-61           rsvd_1.0.5            spatstat.utils_2.3-1 
# [34] cachem_1.0.6          assertthat_0.2.1      promises_1.2.0.1     
# [37] scales_1.2.0          googlesheets4_1.0.0   rgeos_0.5-9          
# [40] gtable_0.3.0          globals_0.15.1        processx_3.7.0       
# [43] goftest_1.2-3         rlang_1.0.4           systemfonts_1.0.4    
# [46] GlobalOptions_0.1.2   splines_4.2.0         lazyeval_0.2.2       
# [49] gargle_1.2.0          spatstat.geom_2.4-0   broom_1.0.0          
# [52] BiocManager_1.30.18   reshape2_1.4.4        abind_1.4-5          
# [55] modelr_0.1.8          backports_1.4.1       httpuv_1.6.5         
# [58] tools_4.2.0           usethis_2.1.6         ellipsis_0.3.2       
# [61] spatstat.core_2.4-4   jquerylib_0.1.4       BiocGenerics_0.42.0  
# [64] sessioninfo_1.2.2     ggridges_0.5.3        plyr_1.8.7           
# [67] ps_1.7.1              prettyunits_1.1.1     rpart_4.1.16         
# [70] deldir_1.0-6          pbapply_1.5-0         GetoptLong_1.0.5     
# [73] urlchecker_1.0.1      S4Vectors_0.34.0      zoo_1.8-10           
# [76] haven_2.5.0           ggrepel_0.9.1         cluster_2.1.3        
# [79] fs_1.5.2              magrittr_2.0.3        RSpectra_0.16-1      
# [82] data.table_1.14.2     scattermore_0.8       lmtest_0.9-40        
# [85] reprex_2.0.1          RANN_2.6.1            googledrive_2.0.0    
# [88] fitdistrplus_1.1-8    matrixStats_0.62.0    pkgload_1.3.0        
# [91] hms_1.1.1             mime_0.12             xtable_1.8-4         
# [94] readxl_1.4.0          IRanges_2.30.0        gridExtra_2.3        
# [97] shape_1.4.6           compiler_4.2.0        KernSmooth_2.23-20   
# [100] crayon_1.5.1          R.oo_1.25.0           htmltools_0.5.3      
# [103] mgcv_1.8-40           later_1.3.0           tzdb_0.3.0           
# [106] lubridate_1.8.0       DBI_1.1.3             dbplyr_2.2.1         
# [109] MASS_7.3-58           Matrix_1.4-1          cli_3.3.0            
# [112] R.methodsS3_1.8.2     parallel_4.2.0        igraph_1.3.4         
# [115] pkgconfig_2.0.3       plotly_4.10.0         spatstat.sparse_2.1-1
# [118] xml2_1.3.3            foreach_1.5.2         bslib_0.4.0          
# [121] rvest_1.0.2           callr_3.7.1           digest_0.6.29        
# [124] RcppAnnoy_0.0.19      spatstat.data_2.2-0   cellranger_1.1.0     
# [127] leiden_0.4.2          uwot_0.1.11           curl_4.3.2           
# [130] rjson_0.2.21          lifecycle_1.0.1       nlme_3.1-158         
# [133] jsonlite_1.8.0        viridisLite_0.4.0     fansi_1.0.3          
# [136] pillar_1.8.0          lattice_0.20-45       fastmap_1.1.0        
# [139] httr_1.4.3            pkgbuild_1.3.1        survival_3.3-1       
# [142] glue_1.6.2            remotes_2.4.2         png_0.1-7            
# [145] iterators_1.0.14      stringi_1.7.8         sass_0.4.2           
# [148] profvis_0.3.7         textshaping_0.3.6     memoise_2.0.1        
# [151] irlba_2.3.5           future.apply_1.9.0   
```
